### DISCUSSION TITLE
Astro component context API

### DISCUSSION THREAD

**Discussion by pilcrowonpaper** [withastro/roadmap#445] [Proposal]
## Overview

Add context API to Astro components

```ts
Astro.setContext(key, value);
Astro.getContext<Value>(key);
```

These contexts are available to all child Astro components of the component. I've thought about making it available to framework components as well, but I could not find a way that works with all major frameworks.

## Usage

Some data, such as language preference, needs to used by all components. However, the only way to share data between parent and child components is to use props, leading to prop drilling. 

## API

The API is based on Svelte's context API. There's also Solid/React version of it but I'm not sure what advantage it provides, since it requires to set a provider wrapper and you need to import the context.

### Types

```ts
const setContext: (key: string | number | symbol, value: any) => void;
const getContext: <V extends any>(key: string | number | symbol) => V | null;
```

**Comment by mandar1jn** [comment]
How would this handle circular references?

**Reply by pilcrowonpaper** [reply]
I don't follow, can you give an example?

**Comment by benoitv-code** [comment]
FYI - solid-start just implemented something like this under the name `locals`: https://github.com/solidjs/solid-start/commit/121d45d27be519bad3e2fc9f9c895dc64f7a0d3d, https://github.com/solidjs/solid-start/issues/642

**Reply by pilcrowonpaper** [reply]
`locals` is for sending data from middleware to request handlers, not for sharing data to child components. Typing an object is going to become messy compared to using getters and setters.

**Reply by benoitv-code** [reply]
IMO the `.locals` would be similar to getting access to context within `.astro` components, while their [useServerContext](https://start.solidjs.com/api/useServerEvent) would be similar to getting access to context within jsx components.

**Comment by lloydjatkinson** [comment]
I keep hitting many use cases for this, all the time. Like, `Tab` and `TabItem` or `ButtonGroup` and `Button`. Some components are strongly coupled, and the context idea is pretty common in other frameworks.

**Comment by Fryuni** [comment]
This would be mostly a simple implementation. Each time you call a component renderer it appends its own context map that it can modify on the first assignment. Any context read walks backwards and read the context from before from the first map that provides it. Two independent instances of the same component will create different contexts.

This works fine for circular dependencies, it will be infinitely deep only if the page is already infinitely deep, which breaks anyway.

We could provide the available context for the UI frameworks and then it is up to each framework how to make it available on their own context. I know it is easily doable for React, Preact and Solid, but I'd have to check for the others. It could even be provided to the client side, but it would have to be configurable in some way to prevent exposing context values containing secrets.

As for the API, I don't like the signature for the `getContext`. That type parameter makes this call the same as `as unknown as T`. It provides no type safety whatsoever.
If we allow arbitrary keys, I'd prefer if the signature for reading was `getContext(key): unknown` so the caller either has to:
- cast explicitly and assume the risk of the value not being of the correct type
- check if the value is of the correct type.

An alternative API would be similar to React's where first a reference has to be created with the type and then it can be tied to both the getter and setter like so:

```ts
// Exported from somewhere, name TBD
const createAstroContext: <T>() => AstroContextRef<T>

interface AstroGlobal {
  setContext<T>(key: AstroContextRef<T>, value: T) => void;
  getContext<T>(key: AstroContextRef<T>) => T | null;
}
```

This would be type safe and used like this:
```ts
// ButtonGroup.astro
import {createAstroContext} from 'astro';

export const myButtonContext = createAstroContext<{color: string}>();

Astro.setContext(myButtonContext, {color: Astro.props.color});

// Button.astro
import {myButtonContext} from './ButtonGroup.astro';

const {color} = Astro.getContext(myButtonContext);
```
